HPCC-13173 Constant folding should not call plugin functions

richardkchapman opened the pull request:
Only specifically whitelisted functions should be constant-folded.

Note that this does not prevent people from declaring their own services and
marking the functions in them as ,fold, so it does not close the security
loophole just yet - that will require some work on the code signing issue
HPCC-14185 - but this is a necessary step towards it (and a useful feature
in its own right).

Signed-off-by: Richard Chapman rchapman@hpccsystems.com

richardkchapman committed:
HPCC-13173 Constant folding should not call plugin functions

Only specifically whitelisted functions should be constant-folded.

Note that this does not prevent people from declaring their own services and
marking the functions in them as ,fold, so it does not close the security
loophole just yet - that will require some work on the code signing issue
HPCC-14185 - but this is a necessary step towards it (and a useful feature
in its own right).

Added ,NOFOLD attribute to override ,FOLD on individual functions within a
service.

Marked appropriate existing service functions foldable.

Refactored the external function call folding code to work around a compiler
bug on OSX.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman commented:
@ghalliday Please review

richardkchapman commented on ecl/hql/hqlatoms.cpp:
```diff
@@ -178,7 +178,7 @@ IAtom * fixedAtom;
 IAtom * flagAtom;
 IAtom * flagsAtom;
 IAtom * flatAtom;
-IAtom * _folded_Atom;
```
This appeared to be unused so I removed it.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented:
@richardkchapman I am surprised that this didn't require a change in hqlgram2.cpp checkServiceDefinition.  I suspect if you define a service function in a non-plugin this will generate a warning.
I'll run a full regression test and see what problems it causes.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented:
@richardkchapman running on the regression suites I get the following problems:

The following fail in testing/regress
stringlib3.ecl 
unicodelib3
teststdlibrary

Some in ecl/regress
fold2
stringlib1

And various queries aren't as efficient, mainly because some of the following calls are no longer folded:
- compiler helper functions in hqlcppsys
- metaphone
- datalib
- possibly some unicode

richardkchapman commented:
@ghalliday Pushed some fixes. There is also an LN PR.

richardkchapman commented on ecl/hql/hqlfold.cpp:
```diff
@@ -557,73 +557,84 @@ static bool isSimpleComparisonArg(IHqlExpression * expr)
  *       functions need to be declared with extern "C".
  *********************************************************/
  //MORE: This function never unloads the plugin dll - this may cause problems in the long run.
-IValue * foldExternalCall(IHqlExpression* expr, unsigned foldOptions, ITemplateContext *templateContext) 
```
I had to split this function to avoid a compiler/glib bug on osx - the embedded asm was interfering with the ability to unwind stack frames on throw.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-trunk0.el6.x86_64.rpm
HPCC Start: OK

| setup | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| setup | Pass | Pass | Pass |
| setup_fetch | Pass | Pass | Pass |
| setupdict | Pass | Pass | Pass |
| setupsearchindex | Pass | Pass | Pass |
| setupsq | Pass | Pass | Pass |
| setupwordindex | Pass | Pass | Pass |
| setupxml | Pass | Pass | Pass |
 

| test | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| stringlib3 | Fail | Fail | Fail |
 
```xml
stringlib3 on hthor failed with:
--- stringlib3:xml
+++ stringlib3:xml
@@ -20,47 +20,47 @@
<Row><Result_7>StringLib:EbcdicStringFind(const ebcdic string src, const ebcdic string tofind , unsigned4 instance )</Result_7></Row>
</Dataset>
<Dataset name=Result 8>
- <Row><Result_10>Str:Extract(const string src, unsigned4 instance)</Result_10></Row>
+ <Row><Result_8>Str:Extract(const string src, unsigned4 instance)</Result_8></Row>
</Dataset>
<Dataset name=Result 9>
- <Row><Result_11>StringLib:GetDateYYYYMMDD()</Result_11></Row>
+ <Row><Result_9>StringLib:GetDateYYYYMMDD()</Result_9></Row>
</Dataset>
<Dataset name=Result 10>
- <Row><Result_12>StringLib:GetBuildInfo()</Result_12></Row>
+ <Row><Result_10>StringLib:GetBuildInfo()</Result_10></Row>
</Dataset>
<Dataset name=Result 11>
- <Row><Result_13>StringLib:Data2String(const data src)</Result_13></Row>
+ <Row><Result_11>StringLib:Data2String(const data src)</Result_11></Row>
</Dataset>
<Dataset name=Result 12>
- <Row><Result_14>StringLib:String2Data(const string src)</Result_14></Row>
+ <Row><Result_12>StringLib:String2Data(const string src)</Result_12></Row>
</Dataset>
<Dataset name=Result 13>
- <Row><Result_15>Str:ToLowerCase(const string src)</Result_15></Row>
+ <Row><Result_13>Str:ToLowerCase(const string src)</Result_13></Row>
</Dataset>
<Dataset name=Result 14>
- <Row><Result_16>Str:ToUpperCase(const string src)</Result_16></Row>
+ <Row><Result_14>Str:ToUpperCase(const string src)</Result_14></Row>
</Dataset>
<Dataset name=Result 15>
- <Row><Result_17>Str:ToCapitalCase(const string src)</Result_17></Row>
+ <Row><Result_15>Str:ToCapitalCase(const string src)</Result_15></Row>
</Dataset>
<Dataset name=Result 16>
- <Row><Result_18>Str:CompareIgnoreCase(const string src1, string src2)</Result_18></Row>
+ <Row><Result_16>Str:CompareIgnoreCase(const string src1, string src2)</Result_16></Row>
</Dataset>
<Dataset name=Result 17>
- <Row><Result_19>Str:Reverse(const string src)</Result_19></Row>
+ <Row><Result_17>Str:Reverse(const string src)</Result_17></Row>
</Dataset>
<Dataset name=Result 18>
- <Row><Result_20>Str:FindReplace(const string src, const string stok, const string rtok)</Result_20></Row>
+ <Row><Result_18>Str:FindReplace(const string src, const string stok, const string rtok)</Result_18></Row>
</Dataset>
<Dataset name=Result 19>
- <Row><Result_21>Str:CleanSpaces(const string src)</Result_21></Row>
+ <Row><Result_19>Str:CleanSpaces(const string src)</Result_19></Row>
</Dataset>
<Dataset name=Result 20>
- <Row><Result_22>Str:WildMatch(const string src, const string _pattern, boolean _noCase)</Result_22></Row>
+ <Row><Result_20>FAILED Str:WildMatch(const string src, const string _pattern, boolean _noCase)</Result_20></Row>
</Dataset>
<Dataset name=Result 21>
- <Row><Result_23>Str:Contains(const string src, const string _pattern, boolean _noCase)</Result_23></Row>
+ <Row><Result_21>FAILED Str:Contains(const string src, const string _pattern, boolean _noCase)</Result_21></Row>
</Dataset>
<Dataset name=Result 22>
- <Row><Result_24>Str:ExtractMultiple(commastring, bitmap)</Result_24></Row>
+ <Row><Result_22>Str:ExtractMultiple(commastring, bitmap)</Result_22></Row>
</Dataset>
stringlib3 on thor failed with:
--- stringlib3:xml
+++ stringlib3:xml
@@ -20,47 +20,47 @@
<Row><Result_7>StringLib:EbcdicStringFind(const ebcdic string src, const ebcdic string tofind , unsigned4 instance )</Result_7></Row>
</Dataset>
<Dataset name=Result 8>
- <Row><Result_10>Str:Extract(const string src, unsigned4 instance)</Result_10></Row>
+ <Row><Result_8>Str:Extract(const string src, unsigned4 instance)</Result_8></Row>
</Dataset>
<Dataset name=Result 9>
- <Row><Result_11>StringLib:GetDateYYYYMMDD()</Result_11></Row>
+ <Row><Result_9>StringLib:GetDateYYYYMMDD()</Result_9></Row>
</Dataset>
<Dataset name=Result 10>
- <Row><Result_12>StringLib:GetBuildInfo()</Result_12></Row>
+ <Row><Result_10>StringLib:GetBuildInfo()</Result_10></Row>
</Dataset>
<Dataset name=Result 11>
- <Row><Result_13>StringLib:Data2String(const data src)</Result_13></Row>
+ <Row><Result_11>StringLib:Data2String(const data src)</Result_11></Row>
</Dataset>
<Dataset name=Result 12>
- <Row><Result_14>StringLib:String2Data(const string src)</Result_14></Row>
+ <Row><Result_12>StringLib:String2Data(const string src)</Result_12></Row>
</Dataset>
<Dataset name=Result 13>
- <Row><Result_15>Str:ToLowerCase(const string src)</Result_15></Row>
+ <Row><Result_13>Str:ToLowerCase(const string src)</Result_13></Row>
</Dataset>
<Dataset name=Result 14>
- <Row><Result_16>Str:ToUpperCase(const string src)</Result_16></Row>
+ <Row><Result_14>Str:ToUpperCase(const string src)</Result_14></Row>
</Dataset>
<Dataset name=Result 15>
- <Row><Result_17>Str:ToCapitalCase(const string src)</Result_17></Row>
+ <Row><Result_15>Str:ToCapitalCase(const string src)</Result_15></Row>
</Dataset>
<Dataset name=Result 16>
- <Row><Result_18>Str:CompareIgnoreCase(const string src1, string src2)</Result_18></Row>
+ <Row><Result_16>Str:CompareIgnoreCase(const string src1, string src2)</Result_16></Row>
</Dataset>
<Dataset name=Result 17>
- <Row><Result_19>Str:Reverse(const string src)</Result_19></Row>
+ <Row><Result_17>Str:Reverse(const string src)</Result_17></Row>
</Dataset>
<Dataset name=Result 18>
- <Row><Result_20>Str:FindReplace(const string src, const string stok, const string rtok)</Result_20></Row>
+ <Row><Result_18>Str:FindReplace(const string src, const string stok, const string rtok)</Result_18></Row>
</Dataset>
<Dataset name=Result 19>
- <Row><Result_21>Str:CleanSpaces(const string src)</Result_21></Row>
+ <Row><Result_19>Str:CleanSpaces(const string src)</Result_19></Row>
</Dataset>
<Dataset name=Result 20>
- <Row><Result_22>Str:WildMatch(const string src, const string _pattern, boolean _noCase)</Result_22></Row>
+ <Row><Result_20>FAILED Str:WildMatch(const string src, const string _pattern, boolean _noCase)</Result_20></Row>
</Dataset>
<Dataset name=Result 21>
- <Row><Result_23>Str:Contains(const string src, const string _pattern, boolean _noCase)</Result_23></Row>
+ <Row><Result_21>FAILED Str:Contains(const string src, const string _pattern, boolean _noCase)</Result_21></Row>
</Dataset>
<Dataset name=Result 22>
- <Row><Result_24>Str:ExtractMultiple(commastring, bitmap)</Result_24></Row>
+ <Row><Result_22>Str:ExtractMultiple(commastring, bitmap)</Result_22></Row>
</Dataset>
stringlib3 on roxie failed with:
--- stringlib3:xml
+++ stringlib3:xml
@@ -20,47 +20,47 @@
<Row><Result_7>StringLib:EbcdicStringFind(const ebcdic string src, const ebcdic string tofind , unsigned4 instance )</Result_7></Row>
</Dataset>
<Dataset name=Result 8>
- <Row><Result_10>Str:Extract(const string src, unsigned4 instance)</Result_10></Row>
+ <Row><Result_8>Str:Extract(const string src, unsigned4 instance)</Result_8></Row>
</Dataset>
<Dataset name=Result 9>
- <Row><Result_11>StringLib:GetDateYYYYMMDD()</Result_11></Row>
+ <Row><Result_9>StringLib:GetDateYYYYMMDD()</Result_9></Row>
</Dataset>
<Dataset name=Result 10>
- <Row><Result_12>StringLib:GetBuildInfo()</Result_12></Row>
+ <Row><Result_10>StringLib:GetBuildInfo()</Result_10></Row>
</Dataset>
<Dataset name=Result 11>
- <Row><Result_13>StringLib:Data2String(const data src)</Result_13></Row>
+ <Row><Result_11>StringLib:Data2String(const data src)</Result_11></Row>
</Dataset>
<Dataset name=Result 12>
- <Row><Result_14>StringLib:String2Data(const string src)</Result_14></Row>
+ <Row><Result_12>StringLib:String2Data(const string src)</Result_12></Row>
</Dataset>
<Dataset name=Result 13>
- <Row><Result_15>Str:ToLowerCase(const string src)</Result_15></Row>
+ <Row><Result_13>Str:ToLowerCase(const string src)</Result_13></Row>
</Dataset>
<Dataset name=Result 14>
- <Row><Result_16>Str:ToUpperCase(const string src)</Result_16></Row>
+ <Row><Result_14>Str:ToUpperCase(const string src)</Result_14></Row>
</Dataset>
<Dataset name=Result 15>
- <Row><Result_17>Str:ToCapitalCase(const string src)</Result_17></Row>
+ <Row><Result_15>Str:ToCapitalCase(const string src)</Result_15></Row>
</Dataset>
<Dataset name=Result 16>
- <Row><Result_18>Str:CompareIgnoreCase(const string src1, string src2)</Result_18></Row>
+ <Row><Result_16>Str:CompareIgnoreCase(const string src1, string src2)</Result_16></Row>
</Dataset>
<Dataset name=Result 17>
- <Row><Result_19>Str:Reverse(const string src)</Result_19></Row>
+ <Row><Result_17>Str:Reverse(const string src)</Result_17></Row>
</Dataset>
<Dataset name=Result 18>
- <Row><Result_20>Str:FindReplace(const string src, const string stok, const string rtok)</Result_20></Row>
+ <Row><Result_18>Str:FindReplace(const string src, const string stok, const string rtok)</Result_18></Row>
</Dataset>
<Dataset name=Result 19>
- <Row><Result_21>Str:CleanSpaces(const string src)</Result_21></Row>
+ <Row><Result_19>Str:CleanSpaces(const string src)</Result_19></Row>
</Dataset>
<Dataset name=Result 20>
- <Row><Result_22>Str:WildMatch(const string src, const string _pattern, boolean _noCase)</Result_22></Row>
+ <Row><Result_20>FAILED Str:WildMatch(const string src, const string _pattern, boolean _noCase)</Result_20></Row>
</Dataset>
<Dataset name=Result 21>
- <Row><Result_23>Str:Contains(const string src, const string _pattern, boolean _noCase)</Result_23></Row>
+ <Row><Result_21>FAILED Str:Contains(const string src, const string _pattern, boolean _noCase)</Result_21></Row>
</Dataset>
<Dataset name=Result 22>
- <Row><Result_24>Str:ExtractMultiple(commastring, bitmap)</Result_24></Row>
+ <Row><Result_22>Str:ExtractMultiple(commastring, bitmap)</Result_22></Row>
</Dataset>

```
 
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-trunk0.el6.x86_64.rpm
HPCC Start: OK

| setup | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| setup | Pass | Pass | Pass |
| setup_fetch | Pass | Pass | Pass |
| setupdict | Pass | Pass | Pass |
| setupsearchindex | Pass | Pass | Pass |
| setupsq | Pass | Pass | Pass |
| setupwordindex | Pass | Pass | Pass |
| setupxml | Pass | Pass | Pass |
 

| test | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| stringlib3 | Fail | Fail | Fail |
 
```xml
stringlib3 on hthor failed with:
--- stringlib3:xml
+++ stringlib3:xml
@@ -56,10 +56,10 @@
<Row><Result_19>Str:CleanSpaces(const string src)</Result_19></Row>
</Dataset>
<Dataset name=Result 20>
- <Row><Result_20>Str:WildMatch(const string src, const string _pattern, boolean _noCase)</Result_20></Row>
+ <Row><Result_20>FAILED Str:WildMatch(const string src, const string _pattern, boolean _noCase)</Result_20></Row>
</Dataset>
<Dataset name=Result 21>
- <Row><Result_21>Str:Contains(const string src, const string _pattern, boolean _noCase)</Result_21></Row>
+ <Row><Result_21>FAILED Str:Contains(const string src, const string _pattern, boolean _noCase)</Result_21></Row>
</Dataset>
<Dataset name=Result 22>
<Row><Result_22>Str:ExtractMultiple(commastring, bitmap)</Result_22></Row>
stringlib3 on thor failed with:
--- stringlib3:xml
+++ stringlib3:xml
@@ -56,10 +56,10 @@
<Row><Result_19>Str:CleanSpaces(const string src)</Result_19></Row>
</Dataset>
<Dataset name=Result 20>
- <Row><Result_20>Str:WildMatch(const string src, const string _pattern, boolean _noCase)</Result_20></Row>
+ <Row><Result_20>FAILED Str:WildMatch(const string src, const string _pattern, boolean _noCase)</Result_20></Row>
</Dataset>
<Dataset name=Result 21>
- <Row><Result_21>Str:Contains(const string src, const string _pattern, boolean _noCase)</Result_21></Row>
+ <Row><Result_21>FAILED Str:Contains(const string src, const string _pattern, boolean _noCase)</Result_21></Row>
</Dataset>
<Dataset name=Result 22>
<Row><Result_22>Str:ExtractMultiple(commastring, bitmap)</Result_22></Row>
stringlib3 on roxie failed with:
--- stringlib3:xml
+++ stringlib3:xml
@@ -56,10 +56,10 @@
<Row><Result_19>Str:CleanSpaces(const string src)</Result_19></Row>
</Dataset>
<Dataset name=Result 20>
- <Row><Result_20>Str:WildMatch(const string src, const string _pattern, boolean _noCase)</Result_20></Row>
+ <Row><Result_20>FAILED Str:WildMatch(const string src, const string _pattern, boolean _noCase)</Result_20></Row>
</Dataset>
<Dataset name=Result 21>
- <Row><Result_21>Str:Contains(const string src, const string _pattern, boolean _noCase)</Result_21></Row>
+ <Row><Result_21>FAILED Str:Contains(const string src, const string _pattern, boolean _noCase)</Result_21></Row>
</Dataset>
<Dataset name=Result 22>
<Row><Result_22>Str:ExtractMultiple(commastring, bitmap)</Result_22></Row>

```
 
HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented:
@richardkchapman still a difference with a couple of the results.

Also, shoult some of the functions in timelib.cpp be constant folded?

richardkchapman commented:
I'll look at timelib.cpp.

Why would the smoketest results be different from what I get when I run the regression suite?

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-trunk0.el6.x86_64.rpm
HPCC Start: OK

| setup | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| setup | Pass | Pass | Pass |
| setup_fetch | Pass | Pass | Pass |
| setupdict | Pass | Pass | Pass |
| setupsearchindex | Pass | Pass | Pass |
| setupsq | Pass | Pass | Pass |
| setupwordindex | Pass | Pass | Pass |
| setupxml | Pass | Pass | Pass |
 

| test | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| stringlib3 | Fail | Fail | Fail |
 
```xml
stringlib3 on hthor failed with:
--- stringlib3:xml
+++ stringlib3:xml
@@ -56,10 +56,10 @@
<Row><Result_19>Str:CleanSpaces(const string src)</Result_19></Row>
</Dataset>
<Dataset name=Result 20>
- <Row><Result_20>Str:WildMatch(const string src, const string _pattern, boolean _noCase)</Result_20></Row>
+ <Row><Result_20>FAILED Str:WildMatch(const string src, const string _pattern, boolean _noCase)</Result_20></Row>
</Dataset>
<Dataset name=Result 21>
- <Row><Result_21>Str:Contains(const string src, const string _pattern, boolean _noCase)</Result_21></Row>
+ <Row><Result_21>FAILED Str:Contains(const string src, const string _pattern, boolean _noCase)</Result_21></Row>
</Dataset>
<Dataset name=Result 22>
<Row><Result_22>Str:ExtractMultiple(commastring, bitmap)</Result_22></Row>
stringlib3 on thor failed with:
--- stringlib3:xml
+++ stringlib3:xml
@@ -56,10 +56,10 @@
<Row><Result_19>Str:CleanSpaces(const string src)</Result_19></Row>
</Dataset>
<Dataset name=Result 20>
- <Row><Result_20>Str:WildMatch(const string src, const string _pattern, boolean _noCase)</Result_20></Row>
+ <Row><Result_20>FAILED Str:WildMatch(const string src, const string _pattern, boolean _noCase)</Result_20></Row>
</Dataset>
<Dataset name=Result 21>
- <Row><Result_21>Str:Contains(const string src, const string _pattern, boolean _noCase)</Result_21></Row>
+ <Row><Result_21>FAILED Str:Contains(const string src, const string _pattern, boolean _noCase)</Result_21></Row>
</Dataset>
<Dataset name=Result 22>
<Row><Result_22>Str:ExtractMultiple(commastring, bitmap)</Result_22></Row>
stringlib3 on roxie failed with:
--- stringlib3:xml
+++ stringlib3:xml
@@ -56,10 +56,10 @@
<Row><Result_19>Str:CleanSpaces(const string src)</Result_19></Row>
</Dataset>
<Dataset name=Result 20>
- <Row><Result_20>Str:WildMatch(const string src, const string _pattern, boolean _noCase)</Result_20></Row>
+ <Row><Result_20>FAILED Str:WildMatch(const string src, const string _pattern, boolean _noCase)</Result_20></Row>
</Dataset>
<Dataset name=Result 21>
- <Row><Result_21>Str:Contains(const string src, const string _pattern, boolean _noCase)</Result_21></Row>
+ <Row><Result_21>FAILED Str:Contains(const string src, const string _pattern, boolean _noCase)</Result_21></Row>
</Dataset>
<Dataset name=Result 22>
<Row><Result_22>Str:ExtractMultiple(commastring, bitmap)</Result_22></Row>

```
 
HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman committed:
HPCC-14242 Constant-folding of external function calls could give wrong answer

Revealed only after refactoring the code in foldExternalCall (earlier in this
branch), the asm sections marshalling parameters were not correctly indicating
their clobber lists, which with some variants of gcc and in release-mode only
could result in the folded result being incorrect.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-trunk0.el6.x86_64.rpm
HPCC Start: OK

| setup | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| setup | Pass | Pass | Pass |
| setup_fetch | Pass | Pass | Pass |
| setupdict | Pass | Pass | Pass |
| setupsearchindex | Pass | Pass | Pass |
| setupsq | Pass | Pass | Pass |
| setupwordindex | Pass | Pass | Pass |
| setupxml | Pass | Pass | Pass |
 

| test | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| stringlib3 | Pass | Pass | Pass |
 
HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented:
@richardkchapman do you want to squash before I merge?

richardkchapman commented:
@ghalliday I have squashed. I kept the missing clobbers as a separate commit as it's fixing a different issue really

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-trunk0.el6.x86_64.rpm
HPCC Start: OK

| setup | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| setup | Pass | Pass | Pass |
| setup_fetch | Pass | Pass | Pass |
| setupdict | Pass | Pass | Pass |
| setupsearchindex | Pass | Pass | Pass |
| setupsq | Pass | Pass | Pass |
| setupwordindex | Pass | Pass | Pass |
| setupxml | Pass | Pass | Pass |
 

| test | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| stringlib3 | Pass | Pass | Pass |
 
HPCC Stop: OK
HPCC Uninstall: OK
